agent/consul/acl_replication_test.go

```diff
@@ -126,6 +126,14 @@ func TestACLReplication_diffACLPolicies(t *testing.T) {
 
 func TestACLReplication_diffACLTokens(t *testing.T) {
 	local := structs.ACLTokens{
+		// just after a 1.3->1.4 upgrade this will happen in the remote
+		&structs.ACLToken{
+			AccessorID:  "",
+			SecretID:    "5128289f-c22c-4d32-936e-7662443f1a55",
+			Description: "token0 - old and not yet upgraded",
+			Hash:        []byte{1, 2, 3, 4},
```

Comment:
I think it's a separate issue, but I'm interested in whether it's possible for a legacy token as seen from the secondary doing the diff to have `Hash: nil`. I've certainly seen in tests when adding legacy tokens that we don't always calculate the hash before storing them in memory, but it's quite possible we always do when deserialising from the RPC or similar.

It's mostly unrelated to the change here, but given you've been deep in this, do you have a quick way to answer that?

Reply:
No Hash _and_ no AccessorID?

Reply:
So I tried this locally:
```
diff --git a/agent/consul/acl_replication.go b/agent/consul/acl_replication.go
index 6d469c5cb..9ee099cd0 100644
--- a/agent/consul/acl_replication.go
+++ b/agent/consul/acl_replication.go
@@ -197,6 +197,12 @@ func diffACLTokens(local structs.ACLTokens, remote structs.ACLTokenListStubs, la
                        remoteIdx += 1
                        continue
                }
+               if len(remote[remoteIdx].Hash) == 0 {
+                       panic(fmt.Sprintf("found remote token during replication without a hash value: %+v", remote[remoteIdx].Hash))
+               }
+               if len(local[localIdx].Hash) == 0 {
+                       panic(fmt.Sprintf("found local token during replication without a hash value: %+v", local[localIdx].Hash))
+               }
                if local[localIdx].AccessorID == remote[remoteIdx].AccessorID {
                        // policy is in both the local and remote state - need to check raft indices and Hash
                        if remote[remoteIdx].ModifyIndex > lastRemoteIndex && !bytes.Equal(remote[remoteIdx].Hash, local[localIdx].Hash) {
```

And did the 2-cluster upgrade dance while creating/using/destroying tokens to upgrade secondary, then primary dcs to 1.4.

The secondary got this lovely panic triggered:

```
    2018/11/07 01:47:31 [DEBUG] http: Request PUT /v1/kv/stuff (3.644234ms) from=10.1.0.1:53692
    2018/11/07 01:47:31 [DEBUG] memberlist: Initiating push/pull sync with: 10.1.2.14:8301
    2018/11/07 01:47:32 [DEBUG] acl: finished fetching remote tokens: 34
    2018/11/07 01:47:32 [DEBUG] acl: token replication - local: 33, remote: 34
panic: found remote token during replication without a hash value: []

goroutine 490 [running]:
github.com/hashicorp/consul/agent/consul.diffACLTokens(0xc0009a8a00, 0x21, 0x40, 0xc0002fb440, 0x22, 0x22, 0xca, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/hashicorp/consul/agent/consul/acl_replication.go:201 +0x98b
github.com/hashicorp/consul/agent/consul.(*Server).replicateACLTokens(0xc000291600, 0xca, 0x2c381a0, 0xc000680d00, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/hashicorp/consul/agent/consul/acl_replication.go:487 +0x52b
github.com/hashicorp/consul/agent/consul.(*Server).startACLReplication.func2(0xc000291600, 0x2c381a0, 0xc000680d00)
	/go/src/github.com/hashicorp/consul/agent/consul/leader.go:784 +0x138
created by github.com/hashicorp/consul/agent/consul.(*Server).startACLReplication
	/go/src/github.com/hashicorp/consul/agent/consul/leader.go:771 +0x1af
```

Reply:
Hmm so it is possible for (legacy?) tokens to make it to the secondary with nil hash. Is that breaking an invariant? Or is the diffing logic somehow tolerant of this?

It may be fine but worth checking over.

Reply:
@mkeeler made a mysterious appearance today and we discussed this. Here is the summary:

 - It is an issue caused by the async "upgrade" code that adds AccessorID being the code path missing a `SetHash` call. (patch below).
 - I hypothesize that the reason we don't actually see errors in our upgrade integration scenarioes (with this PR) is because:
   - The Legacy ACL is diffed based on AccessorID and deemed to be "new" because it's counter part on the secondary doesn't have an AccessorID yet
   - So we don't even bother trying to compare the Hashes
   - We upsert it, which sets Hash locally.
   - Any future change to that legacy token on the primary will go through another code path that does call `SetHash` correctly so any future replication of it will procede with hashes as normal.

So in a sense, with current code, it is actually masked/harmless, however it's another subtlety that we might trip up on later so we should fix.

Thankfully the fix is trivial:

```diff
diff --git a/agent/consul/leader.go b/agent/consul/leader.go
index 71e615801..05a1992ac 100644
--- a/agent/consul/leader.go
+++ b/agent/consul/leader.go
@@ -629,6 +629,8 @@ func (s *Server) startACLUpgrade() {
                                newToken.CreateIndex = token.CreateIndex
                                newToken.ModifyIndex = token.ModifyIndex

+                               newToken.SetHash(true)
+
                                newTokens = append(newTokens, &newToken)
                        }
```

Reply:
Will fix.